docs/design/notifications.md

```diff
@@ -0,0 +1,196 @@
+
+**Why Broker Auth for Broadcast Notifications:**
+
+The broker uses its own authentication credentials (configured at startup) rather than client credentials because broadcast notifications are not tied to any specific client session, must be received even when no clients have made tool calls yet, and allows the broker to maintain a single persistent connection per backend server rather than per client.
+
+#### Client-Specific Notifications
+
+Client-specific notifications are notifications related to a particular client's session or tool execution. Examples include:
+
+- Progress updates for long-running tool calls
+- Tool call completion notifications
+- Elicitation requests (e.g., `elicitation/create`) - These allow backend MCP servers to request user input during tool execution, which is important for confirming destructive actions or obtaining out-of-band authentication URLs
+- Other session-specific state changes
+
+**How Client-Specific Notifications Work:**
+
+```mermaid
+sequenceDiagram
+  participant Client1 as MCP Client 1
+  participant Client2 as MCP Client 2
+  participant Gateway as Gateway/MCP Router/MCP Broker
+  participant Server1 as MCP Server 1
+  participant Server2 as MCP Server 2
+
+  Note over Client1, Server2: Tool Call to Server 1
+  Client1 ->> Gateway: POST /mcp "tool/call"
+  Note over Gateway, Server1: Lazy Init
+  Gateway ->> Server1: POST /mcp "initialize"
+  Gateway ->> Server1: POST /mcp "notifications/initialized"
+  Gateway ->> Server1: GET /mcp
+  Server1 -->> Gateway: SSE connection established
+  Gateway ->> Server1: POST /mcp "tool/call"
+  Server1 ->> Gateway: progress notification (client auth)
+  Gateway ->> Client1: forward notification
+  Server1 ->> Gateway: tool/call response
+  Gateway ->> Client1: tool/call response
+  Gateway -->> Server1: SSE connection closed
+  Note over Client1, Server2: Tool Call to Server 2
+  Note over Gateway, Server2: Lazy Init
+  Gateway ->> Server2: POST /mcp "initialize"
+  Gateway ->> Server2: POST /mcp "notifications/initialized"
+  Gateway ->> Server2: GET /mcp
+  Server2 -->> Gateway: SSE connection established
+  Gateway ->> Server2: POST /mcp "tool/call"
+  Server2 ->> Gateway: progress notification (client auth)
+  Gateway ->> Client1: forward notification
+  Server2 ->> Gateway: tool/call response
+  Gateway ->> Client1: tool/call response
+  Gateway -->> Server2: SSE connection closed
+```
+
+1. **Lazy Session Establishment**: When a client makes a `tools/call` request, the broker performs lazy initialization to establish a backend session to the target MCP server using the client's authentication credentials (passed in the tool call request). This initialization includes:
+   - Sending an `initialize` request to the backend server *only* if a session hasn't already been established from a previous tool call.
+   - Sending a `notifications/initialized` notification
+   - Establishing a GET `/mcp` connection for receiving SSE notifications
```

Comment:
would this get be done with the clients session id? I assume yes?

Reply:
Yes. It has to to client specific